design/design-doc-composition.md

```diff
@@ -0,0 +1,816 @@
+  name: wordpresses.apps.example.org
+spec:
+  # Defines the structural schema and GroupVersionKind of this application. Only
+  # a single API version of the application may exist. Additional fields will be
+  # injected to support composition machinery.
+  crdSpecTemplate:
+    group: apps.example.org
+    version: v1alpha1
+    names:
+      kind: Wordpress
+      listKind: WordpressList
+      plural: wordpresses
+      singular: wordpress
+    validation:
+      openAPIV3Schema:
+        properties:
+          administratorLogin:
+            type: string
+          storageSize:
+            type: int
+          storageType:
+            type: string
+        type: object
+```
+
+When an application developer authors the above `ApplicationDefinition`
+Crossplane will automatically create a corresponding `CustomResourceDefinition`
+allowing application operators to author the below custom resource:
+
+```yaml
+apiVersion: example.org/v1alpha1
+kind: Wordpress
+metadata:
+  namespace: default
+  name: coolblog
+spec:
+  # The schema for the following three fields is defined by the above
+  # ApplicationDefinition.
+  administratorLogin: admin
+  storageSize: 2
+  storageType: SSD
+  # The application object schema is automatically injected into the
+  # CustomResourceDefinition that is created by the ApplicationDefinition that
+  # defines the Wordpress resource.
+  application:
+    # Multiple compositions may potentially satisfy a particular kind of
+    # application. Exactly one composition may be annotated as the default for
+    # a particular kind of application. Each application instance may influence
+    # which composition is used via label selectors. This could be used, for
+    # example, to determine whether a Wordpress application renders to a
+    # KubernetesApplication or to a plain old Kubernetes Deployment.
+    compositionSelector:
+      matchLabels:
+        compute: kubernetes
+        database: mysql
+    # The Wordpress author may explicitly select which composition should be
+    # used by setting the compositionRef. In the majority of cases the author
+    # will ignore this field and it will be set by a controller, similar to the
+    # contemporary classRef field.
+    compositionRef:
+    - name: wordpress-kubernetes-mysql
+status:
+  application:
+    # Each application maintains an array of the resources they compose.
+    # Composed resources are always in the same namespace as the application
+    # resource. Any namespaced resource may be composed; composed resources
+    # model their relationship with the application resource via their
+    # controller reference. The application must maintain this array because
+    # there is currently no user friendly, performant way to discover which
+    # resources (of arbitrary kinds) are controlled by a particular resource per
+    # https://github.com/kubernetes/kubernetes/issues/54498
+    composedRefs:
```

Comment:
Yeah, I was on the fence about this.

Ideally these references would be purely informational - they would exist only to inform consumers of a custom resource of its children. This would require us to generate a deterministic, stable name for each child so that the controller wouldn't need to read back these references - it could just generate the same name for each of its composed children each time.

Realistically I suspect the controller will have to read these back, and that we'd thus orphan our child resources and create a new set if it were lost. So it should probably go in spec. My main concern here is that I don't want to support or encourage the case in which these references are manually set by the author of the CR.